helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentMetadataStore.java

```diff
@@ -28,24 +38,114 @@
  * A placeholder before we have the real assignment metadata store.
  */
 public class AssignmentMetadataStore {
-  private Map<String, ResourceAssignment> _persistGlobalBaseline = new HashMap<>();
-  private Map<String, ResourceAssignment> _persistBestPossibleAssignment = new HashMap<>();
+  private static final String ASSIGNMENT_METADATA_KEY = "ASSIGNMENT_METADATA";
+  private static final String BASELINE_KEY = "BASELINE";
+  private static final String BEST_POSSIBLE_KEY = "BEST_POSSIBLE";
+  private static final ZkSerializer SERIALIZER = new ZNRecordJacksonSerializer();
+
+  private BucketDataAccessor _dataAccessor;
+  private String _baselinePath;
+  private String _bestPossiblePath;
+  private Map<String, ResourceAssignment> _globalBaseline;
```

Comment:
Actually I don't think these 2 private fields are necessary unless you want to have "lazy cache" functionality that "get" from DB once and persist.  But be aware of the approach that the client probably won't always be able to get the latest data

Reply:
Could you re-word your comment? I don't think I'm understanding this clearly.

The reason I am keeping these two private fields is if there are no connection losses or controller switches, there wouldn't need to read everything from ZK since we'd already have it in-memory.

Reply:
"Already have it in-memory" that's the idea of lazy cache: read from DB and keeps in memory. However, it's possible the data in memory is out of date, and you will need to consider when the client needs to refresh its "cached" data. 
To keep things simpler, if performance is not the bottleneck, always read from ZK will guarantee the correct up-to-date data. In this case, these 2 private fields are not necessary.
Also, curious that in the case of the controller switch, the memory is literally gone, isn't it?

Reply:
1. Performance is critical here.
2. For AssignmentMetadataStore, there should be only one writer. The controller. So if we verify if the current manager is the leader, we shall be good.

Reply:
@i3wangyi How could the data in memory be out-of-date? I think there's a gap in understanding somewhere...

Reply:
For example, there's a custom tool that updates the baseline and best possible mapping after this metadata class reads & caches the result from ZK. 
Also, take your exposed 2 methods (`getbasline() and persistBaseline()`). I understand the `persistBaseline` will overwrite the local private `_baseline` in the method body. In the concurrency environment, it will cause the race condition and it creates a read after write consistency issue (the in-memory data will not get updated).

Reply:
Ah I see where the disconnect is: 
`AssignmentMetadataStore` is for WAGED rebalancer use only. That means that no other tool should update the assignment metadata. There would be only one instance of WAGED Rebalancer and this `AssignmentMetadataStore` will be a static object or a singleton, so we do not have to guard against concurrent writes.

Reply:
Got it. But there's no such constraint based on the class design. First, it's a public class; Second, it has a public constructor, meaning everyone could create an instance not subjective to waged rebalancer; I suggest enforcing it by singleton pattern (a factory constructor and private constructors).

Reply:
I think using the factory constructor is a bit of an overkill, but I think your point is valid for making it package private so that only waged rebalancer could use it :)